tests/unittest_nnstreamer-edge.cc

```diff
@@ -47,6 +123,143 @@ _test_edge_event_cb (nns_edge_event_h event_h, void *user_data)
+  pthread_attr_t attr;
+  size_t data_len;
+  void *data;
+  unsigned int i;
+  int ret, port;
+  char *val;
+
+  _td_server = _get_test_data (true);
+  _td_client1 = _get_test_data (false);
+  _td_client2 = _get_test_data (false);
+  port = nns_edge_get_available_port ();
+
+  /* Prepare server (127.0.0.1:port) */
+  val = nns_edge_strdup_printf ("%d", port);
+  nns_edge_create_handle ("temp-server", "temp-topic", &server_h);
+  nns_edge_set_event_callback (server_h, _test_edge_event_cb, _td_server);
+  nns_edge_set_info (server_h, "IP", "127.0.0.1");
+  nns_edge_set_info (server_h, "PORT", val);
+  nns_edge_set_info (server_h, "CAPS", "test server");
+  _td_server->handle = server_h;
+  nns_edge_free (val);
+
+  /* Prepare client */
+  nns_edge_create_handle ("temp-client1", "temp-topic", &client1_h);
+  nns_edge_set_event_callback (client1_h, _test_edge_event_cb, _td_client1);
+  nns_edge_set_info (client1_h, "CAPS", "test client1");
+  _td_client1->handle = client1_h;
+
+  nns_edge_create_handle ("temp-client2", "temp-topic", &client2_h);
+  nns_edge_set_event_callback (client2_h, _test_edge_event_cb, _td_client2);
+  nns_edge_set_info (client2_h, "CAPS", "test client2");
+  _td_client2->handle = client2_h;
+
+  /* Start server/client thread */
+  pthread_attr_init (&attr);
+  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED);
+  pthread_create (&server_thread, &attr, _test_edge_thread, _td_server);
+  pthread_create (&client1_thread, &attr, _test_edge_thread, _td_client1);
+  pthread_create (&client2_thread, &attr, _test_edge_thread, _td_client2);
+  pthread_attr_destroy (&attr);
+
+  /* Wait for server/client thread */
+  do {
+    usleep (20000);
+  } while (!g_main_loop_is_running (_td_server->loop));
+
+  do {
+    usleep (20000);
+  } while (!g_main_loop_is_running (_td_client1->loop));
+
+  do {
+    usleep (20000);
+  } while (!g_main_loop_is_running (_td_client2->loop));
+
+  ret = nns_edge_connect (client1_h, NNS_EDGE_PROTOCOL_TCP, "127.0.0.1", port);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+  usleep (10000);
+  ret = nns_edge_connect (client2_h, NNS_EDGE_PROTOCOL_TCP, "127.0.0.1", port);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  sleep (2);
+
+  /* Send request to server */
+  data_len = 10U * sizeof (unsigned int);
+  data = malloc (data_len);
+  ASSERT_TRUE (data != NULL);
+
+  for (i = 0; i < 10U; i++)
+    ((unsigned int *) data)[i] = i;
+
+  ret = nns_edge_data_create (&data_h);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  ret = nns_edge_data_add (data_h, data, data_len, free);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  for (i = 0; i < 5U; i++) {
+    ret = nns_edge_request (client1_h, data_h);
+    EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+    usleep (10000);
+    ret = nns_edge_request (client2_h, data_h);
+    EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+    usleep (100000);
+  }
+
+  ret = nns_edge_data_destroy (data_h);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  sleep (2);

```

Comment:
it has always been troubling when we assume that the CI/build server will be reasonably fast....

Reply:
Updated - wait for responding from server handle.